---
title: "RNA-seq analysis of TF KD samples"
author: "Borja Gomez"
date: "2022-08-23"
output: html_document
---

##Introduction

Analysis of RNA-seq data from TF KD samples. The samples used here were late transduction of shRNA particles for the KD of LBX1, NHLH1 and NR2F1/2 with their corresponding shScramble controls. In addition, there are also early transduction of shRNA for the KD of NR2F1/2 and the corresponding shScramble controls. Each TF KD had 3 replicates and then corresponding shScrambles were included. To match shTF with shScramble, use the date.  

RNA samples were collected at day 15 of differentiation.

#Version of the different tools used

```{r}
conda 4.11.0
snakemake 6.10.0

STAR 2.7.4a
samtools 1.10
FastQC v0.11.9
FastQ Screen v0.14.0
AdapterRemoval ver. 2.3.1

R version 4.0.0 (2020-04-24)
Ubuntu 20.04 LTS

package version date    source
ggplot2 3.3.2   2020-06-19      CRAN (R 4.0.0)
DESeq2  1.28.1  2020-05-12      Bioconductor
apeglm  1.10.0  2020-04-27      Bioconductor
Rsubread        2.2.2   2020-05-15      Bioconductor

```

#RNA-seq analysis using SnakeMake

For the analysis of the RNA-seq data Snakemake was used (https://github.com/snakemake/snakemake). 

The workflow for the RNA-seq data can be found in https://gitlab.lcsb.uni.lu/aurelien.ginolhac/snakemake-rna-seq 

First, the environment need to be established:

```{r}
#Snakemake is installed via conda
conda update conda
conda install -c conda-forge mamba
mamba create -c conda-forge -c bioconda -n snakemake snakemake
```

A snakemake conda environment was created for the analysis

```{r}
#It is recommended to update snakemake
conda update -c conda-forge -c bioconda -n snakemake snakemake
```

Everything is ready to fecth the workflow template:

```{r}
wget -qO- https://git-r3lab.uni.lu/aurelien.ginolhac/snakemake-rna-seq/-/archive/v0.2.3/snakemake-rna-seq-.tar.gz | tar xfz - --strip-components=1
```

All of this can be run in the conda base, there is no need to enter the snakemake environment.

Preparation of specific aliases:

```{r}
alias smk='conda activate snakemake && module load tools/Singularity'
alias dag='snakemake --dag | dot -Tpdf > dag.pdf'
complete -o bashdefault -C snakemake-bash-completion snakemake
```

Enter the config folder and go the the following file: **config.yaml**. Inside this file, the following modifications were done:

```{r}
#coding=utf-8
# path or URL to sample sheet (TSV format, columns: sample, condition, ...)
samples: config/samples.tsv
# path or URL to sequencing unit sheet (TSV format, columns: sample, unit, fq1, fq2,
# strandedness). Units are technical replicates (e.g. lanes, or resequencing of the
# same biological sample).If the column "strandedness" is present (which is optional),
# can be empty or has one of these values: none, yes or reverse. none is for unstranded
# protocols, yes an reverse follow the nomenclature used in `htseq-count --reverse`
# which is referenced in STAR manual section 7, "Counting number of reads per gene".

units: config/units.tsv

trimming:
  # skip trimming: false or true
  skip: false
  threads: 2
  # adapter removal
  adapter1: AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC
  adapter2: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT
  min_length: 35

ref:
  species: "homo_sapiens"
  build: "GRCh38"
  release: "102"

pca:
  labels:
    # columns of sample sheet to use for PCA
    - condition

diffexp:
  contrasts:
   # contrasts for the deseq2 results method
   # write down EFFECT_vs_CONTROL:
   #  - FULLNAME_CONTROL
   #  - FULLNAME_EFFECT
    KD_vs_Scramble:
      - Scramble
      - KD


params:
  star: " --twopassMode Basic --outSAMtype BAM SortedByCoordinate --limitOutSJcollapsed 1000000 --limitSjdbInsertNsj 1000000 --outFilterMultimapNmax 100 --outFilterMismatchNmax 33 --outFilterMis$
  bowtie_path: "/usr/local/bin/"
  db_bowtie_path: "/scratch/users/aginolhac/FastQ_Screen_Genomes/"

```

In the contrast, KD vs Scramble was defined. However, TF KDs would be analyzed separately and not together, but that can be done later and this just simplify the running of this workflow. 

Create the **units.tsv** file with the following command:

```{r}
ls ~/*fastq.gz | paste - - | awk 'BEGIN{OFS="\t";print "sample\tunit\tfq1\tfq2\tstrandedness"}{id=substr($1,index($1,"/")+1,index($1,"_EFPFU")-7);print id,"1",$1,$2, "reverse"}' | sed 's/\-/_/' > config/units.tsv
```

The file should look as this example:

```{r}
sample  unit    fq1     fq2     strandedness
20210403_D15_Late_LBX1  1       fastq/20210403_D15_Late_LBX1_EFPFU_SB_S69_R1_001.fastq.gz       fastq/20210403_D15_Late_LBX1_EFPFU_SB_S69_R2_001.fastq.gz       reverse
20210403_D15_Late_Scramble      1       fastq/20210403_D15_Late_Scramble_EFPFU_SB_S70_R1_001.fastq.gz   fastq/20210403_D15_Late_Scramble_EFPFU_SB_S70_R2_001.fastq.gz   reverse
20210407_D15_Late_LBX1  1       fastq/20210407_D15_Late_LBX1_EFPFU_SB_S71_R1_001.fastq.gz       fastq/20210407_D15_Late_LBX1_EFPFU_SB_S71_R2_001.fastq.gz       reverse
20210407_D15_Late_Scramble      1       fastq/20210407_D15_Late_Scramble_EFPFU_SB_S72_R1_001.fastq.gz   fastq/20210407_D15_Late_Scramble_EFPFU_SB_S72_R2_001.fastq.gz   reverse
20210413_D15_Late_LBX1  1       fastq/20210413_D15_Late_LBX1_EFPFU_SB_S73_R1_001.fastq.gz       fastq/20210413_D15_Late_LBX1_EFPFU_SB_S73_R2_001.fastq.gz       reverse
20210413_D15_Late_Scramble      1       fastq/20210413_D15_Late_Scramble_EFPFU_SB_S74_R1_001.fastq.gz   fastq/20210413_D15_Late_Scramble_EFPFU_SB_S74_R2_001.fastq.gz   reverse
20210602_D15_Late_NHLH1 1       fastq/20210602_D15_Late_NHLH1_EFPFU_SB_S75_R1_001.fastq.gz      fastq/20210602_D15_Late_NHLH1_EFPFU_SB_S75_R2_001.fastq.gz      reverse
20210602_D15_Late_Scramble      1       fastq/20210602_D15_Late_Scramble_EFPFU_SB_S76_R1_001.fastq.gz   fastq/20210602_D15_Late_Scramble_EFPFU_SB_S76_R2_001.fastq.gz   reverse
20210615_D15_Late_NR2Fs 1       fastq/20210615_D15_Late_NR2Fs_EFPFU_SB_S78_R1_001.fastq.gz      fastq/20210615_D15_Late_NR2Fs_EFPFU_SB_S78_R2_001.fastq.gz      reverse
20210615_D15_Late_Scramble      1       fastq/20210615_D15_Late_Scramble_EFPFU_SB_S79_R1_001.fastq.gz   fastq/20210615_D15_Late_Scramble_EFPFU_SB_S79_R2_001.fastq.gz   reverse
20210615_D15_NR2Fs      1       fastq/20210615_D15_NR2Fs_EFPFU_SB_S80_R1_001.fastq.gz   fastq/20210615_D15_NR2Fs_EFPFU_SB_S80_R2_001.fastq.gz   reverse
20210615_D15_Scramble   1       fastq/20210615_D15_Scramble_EFPFU_SB_S81_R1_001.fastq.gz        fastq/20210615_D15_Scramble_EFPFU_SB_S81_R2_001.fastq.gz        reverse
20210619_D15_Late_NHLH1 1       fastq/20210619_D15_Late_NHLH1_EFPFU_SB_S82_R1_001.fastq.gz      fastq/20210619_D15_Late_NHLH1_EFPFU_SB_S82_R2_001.fastq.gz      reverse
20210619_D15_Late_NR2Fs 1       fastq/20210619_D15_Late_NR2Fs_EFPFU_SB_S83_R1_001.fastq.gz      fastq/20210619_D15_Late_NR2Fs_EFPFU_SB_S83_R2_001.fastq.gz      reverse
20210619_D15_Late_Scramble      1       fastq/20210619_D15_Late_Scramble_EFPFU_SB_S84_R1_001.fastq.gz   fastq/20210619_D15_Late_Scramble_EFPFU_SB_S84_R2_001.fastq.gz   reverse
20210619_D15_NR2Fs      1       fastq/20210619_D15_NR2Fs_EFPFU_SB_S86_R1_001.fastq.gz   fastq/20210619_D15_NR2Fs_EFPFU_SB_S86_R2_001.fastq.gz   reverse
20210619_D15_Scramble   1       fastq/20210619_D15_Scramble_EFPFU_SB_S87_R1_001.fastq.gz        fastq/20210619_D15_Scramble_EFPFU_SB_S87_R2_001.fastq.gz        reverse
20210623_D15_Late_NHLH1 1       fastq/20210623_D15_Late_NHLH1_EFPFU_SB_S88_R1_001.fastq.gz      fastq/20210623_D15_Late_NHLH1_EFPFU_SB_S88_R2_001.fastq.gz      reverse
20210623_D15_Late_NR2Fs 1       fastq/20210623_D15_Late_NR2Fs_EFPFU_SB_S89_R1_001.fastq.gz      fastq/20210623_D15_Late_NR2Fs_EFPFU_SB_S89_R2_001.fastq.gz      reverse
20210623_D15_Late_Scramble      1       fastq/20210623_D15_Late_Scramble_EFPFU_SB_S90_R1_001.fastq.gz   fastq/20210623_D15_Late_Scramble_EFPFU_SB_S90_R2_001.fastq.gz   reverse
20210623_D15_NR2Fs      1       fastq/20210623_D15_NR2Fs_EFPFU_SB_S92_R1_001.fastq.gz   fastq/20210623_D15_NR2Fs_EFPFU_SB_S92_R2_001.fastq.gz   reverse
20210623_D15_Scramble   1       fastq/20210623_D15_Scramble_EFPFU_SB_S93_R1_001.fastq.gz        fastq/20210623_D15_Scramble_EFPFU_SB_S93_R2_001.fastq.gz        reverse

```

Then, **samples.tsv** file needs to be created too. The following command can be used:

```{r}
ls ~/*fastq.gz | paste - - | awk 'BEGIN{OFS="\t";print "sample\tcondition"}{id=substr($1,index($1,"/")+1,index($1,"_EFPFU")-7);print id,id}' | sed 's/[0-9]$//; s/\-/_/g' > config/samples.tsv
```

Inside samples.tsv, condition needs to be specified. Here is an example:

```{r}
sample  condition
20210403_D15_Late_LBX1  KD
20210403_D15_Late_Scramble      Scramble
20210407_D15_Late_LBX1  KD
20210407_D15_Late_Scramble      Scramble
20210413_D15_Late_LBX1  KD
20210413_D15_Late_Scramble      Scramble
20210602_D15_Late_NHLH1 KD
20210602_D15_Late_Scramble      Scramble
20210615_D15_Late_NR2Fs KD
20210615_D15_Late_Scramble      Scramble
20210615_D15_NR2Fs      KD
20210615_D15_Scramble   Scramble
20210619_D15_Late_NHLH1 KD
20210619_D15_Late_NR2Fs KD
20210619_D15_Late_Scramble      Scramble
20210619_D15_NR2Fs      KD
20210619_D15_Scramble   Scramble
20210623_D15_Late_NHLH1 KD
20210623_D15_Late_NR2Fs KD
20210623_D15_Late_Scramble      Scramble
20210623_D15_NR2Fs      KD
20210623_D15_Scramble   Scramble

```

Once those files are defined, snakemake can be run using the following commands:

```{r}
#Remember that a Singularity image would be needed for running this workflow
conda activate snakemake 
srun snakemake --use-singularity --singularity-args "-B /scratch/users/aginolhac:/scratch/users/aginolhac" -j 28
```

Until here the analysis is done and gene counts were obtained. Althought this workflow already runs DESeq2, it uses all the samples and the comparisons should be done per TF and corresponding shScrambles. 

#DESeq2 analysis per TF

Load the necessary libraries;

```{r}
library(tidyverse)
library(DESeq2)
library(knitr)
library(rtracklayer)
library(RColorBrewer)
library(pheatmap)
```

Extract gene counts per TF:

```{r}
fc_hs <- readr::read_rds("~/fc.rds")
rawcounts <- fc_hs$counts
#Here an example for extracting the counts for one of the TF KD conditions

as.data.frame(fc_hs$counts) %>% select(7,8,14,16,20,22) -> rawcounts_TF
```

Create a metadata file

```{r}
#Here an example again
sample_TF <- c("TF_KD", "SCR", "TF_KD", "SCR", "TF_KD", "SCR")

info.data.tf <- data.frame(sample_TF)

rownames(info.data.tf) <- c(colnames(rawcounts_TF))
```

Create the DESeq2 object:

```{r}
dds_TF <- DESeqDataSetFromMatrix(countData = rawcounts_TF,
                              colData = info.data.tf,
                              design = ~ sample_TF)

dds_NHLH1$sample_NHLH1 <- factor(dds_TF$sample_TF, levels = c("SCR", "TF_KD"))
```

For creating a PCA plot:

```{r}
vst_TF <- vst(dds_TF, blind = TRUE)

plotPCA(vst_TF, intgroup = c("sample_TF"))
```

Filtering of low expressed genes:

```{r}
keep <- rowSums(counts(dds_TF)) >= 10

dds_TF <- dds_TF[keep,]
```

Run DESeq2 analysis:

```{r}
dds_TF <- DESeq(dds_TF)
```

For building MAplot and Volcano plots:

```{r}
library(apeglm)

results_LFC_TF <- lfcShrink(dds_TF, coef="sample_TF_TF_KD_vs_SCR", type="apeglm")

#For MA plot
plotMA(results_LFC_TF, ylim=c(-10, 10), alpha = 0.05)

#For Volcano plot
as.data.frame(results_LFC_TF) -> results_LFC_TF

results_LFC_TF %>% mutate(Expression = case_when(log2FoldChange >= 1 & padj < 0.05 ~ "Upregulated",
                   log2FoldChange <= -1 & padj < 0.05 ~ "Downregulated")) -> results_LFC_TF

ggplot(results_LFC_TF, aes(log2FoldChange, -log10(padj))) +
  geom_point(aes(color = Expression), size = 1) +
  theme_light() +
  scale_color_manual(values = c("red", "red")) +
  geom_vline(xintercept = c(-1, 1)) + 
  geom_hline(yintercept = -log10(0.05)) +
  xlim(-5, 5)

```

Results can be exported after filtering for padj for example

#Calculation of FPKM

Rawcount can be extracted from the dds object with the already filtered genes:

```{r}
rawcount_2 <- counts(dds)
dim(rawcount_2)
```
Extract gene length for the genes included in rawcounts. Annotation file needs to be prepared and it is provided by the snakemake workflow, folder **refs**: 

```{r}
gene_length <- fc_hs$annotation %>% select(GeneID, Length) %>% filter(GeneID %in% rownames(rawcount_2))
   
#After extraction of the gene length, we transform the results from bp to kb
gene_length <- gene_length %>% mutate(length_kb = Length/1000)
  
# scaling factor. Here we are obtaining the reads per million for each individual sample. It is row wise. This is indicated using the number 2 in the function. 
scale_factor <- apply(rawcount_2, 2, function(x) sum(x))/1000000
scale_factor
  
# normalize to library size.
as.data.frame(rawcount_2) -> rawcount_2
column_to_rownames(rawcount_2, var = "GeneID") -> rawcount_2
norm_counts <- rawcount_2 %>% sweep(., 2, scale_factor, "/")

  
# fpkm. This is divided again by the gene length. 
rawcount_2 %>% rownames_to_column(var = "GeneID") -> rawcount_2
all_info <- left_join(rawcount_2, gene_length, by = c("GeneID")) 
  
fpkm <- norm_counts %>% sweep(., 1, all_info$length_kb, "/")
fpkm
```

Now fpkm can be used for the plotting of individual genes for example.

#Calculate expression Z-score for heatmaps

```{r}
vsd_TF <- assay(vst(dds_TF))
Z_TF <- t(scale(t(vsd_TF)))
```

Z-score was used for the creation of heatmaps. Genes of interest can be filtered and for creating the heatmap:

```{r}
pheatmap(Z_TF, annotation_col = info.data.tf, show_colnames = FALSE, cluster_cols = FALSE)
```

#Plotting individual genes 

fpkm was used for creating these plots:

```{r}
fpkm$sample <- c("TF_KD", "SCR", "TF_KD", "SCR", "TF_KD", "SCR") 

ggplot(fpkm) + 
  geom_boxplot(aes(x=sample, y= "", fill = sample)) +
  geom_jitter(aes(x=sample, y= ""), alpha = 1) +
  theme_light() +
  labs(title = "",
       x = "",
       y = "FPKM") +
  theme(plot.title = element_text(size = 20, face = "bold", hjust = 0.5), axis.title.y = element_text(face = "bold")) + expand_limits(y = 0) -> Gene_expression

ggsave(plot = Gene_expression, filename = "*.pdf", dpi = 300)
```












