---
title: "RNA-seq analysis. Time course differentiation data"
author: "Borja Gomez"
date: "2022-08-23"
output: html_document
---

##Introduction

Here is contained all the code used for the analysis of the time course differentiation data, including: smNPC, mDAN at days 15/30/50, non-mDAN at days 15/50, and astrocytes at day 65 of differentiation. Each sample contained a total of 3 replicates.

#Version of the different tools used

```{r}
samtools 1.9
Adapter Remval 2.2.2
bwa 0.7.17
fastqc 0.11.7
picard 2.18.13
paleomix 1.2.13.2
STAR 2.5.2b
CMake 3.12.1-intel-2018a
sortmerna 3.0.3
```

#Sample quality control by FastQC and MultiQC 

The FastQC tool (https://github.com/s-andrews/FastQC) was used to determine the quality of the RNA-seq samples obtained:

```{r}
ls *fastq.gz | parallel -j 8 "fastqc --nogroup --outdir=  {}" &
```

Then, in order to obtain all the results in a single report, MultiQC was used (https://multiqc.info/):

```{r}
multiqc #outdir
```

#Adapter trimming using Paleomix

For sequencing adapter removal, PALEOMIX was used (https://paleomix.readthedocs.io/en/stable/):

```{r}
#Create makefile 
paleomix trim_pipeline mkfile > paleo_trim.yaml

#Open the .yaml file to modify it
vim paleo_trim.yaml
```

Once inside the paleo_trim.yaml file, some changes were added to the file plus the path and short name for the fastq files was specified at the end (2 examles are shown):

```{r}
# -*- mode: Yaml; -*-
# Timestamp: 2019-07-18T10:05:18.882102
#
# Default options.
# Can also be specific for a set of samples, libraries, and lanes,
# by including the "Options" hierarchy at the same level as those
# samples, libraries, or lanes below. This does not include
# "Features", which may only be specific globally.
Options:
  # Sequencing platform, see SAM/BAM reference for valid values
  Platform: Illumina
  # Quality offset for Phred scores, either 33 (Sanger/Illumina 1.8+)
  # or 64 (Illumina 1.3+ / 1.5+). For Bowtie2 it is also possible to
  # specify 'Solexa', to handle reads on the Solexa scale. This is
  # used during adapter-trimming and sequence alignment
  QualityOffset: 33
  # Split a lane into multiple entries, one for each (pair of) file(s)
  # found using the search-string specified for a given lane. Each
  # lane is named by adding a number to the end of the given barcode.
  SplitLanesByFilenames: yes
  # Compression format for FASTQ reads; 'gz' for GZip, 'bz2' for BZip2
  CompressionFormat: gz

  # Settings for trimming of reads, see AdapterRemoval man-page
  AdapterRemoval:
     # Adapter sequences, set and uncomment to override defaults
#     --adapter1: AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC
#     --adapter2: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT
     # Some BAM pipeline defaults differ from AR defaults;
     # To override, change these value(s):
 --mm: 3
     --minlength: 25
     # Extra features enabled by default; change 'yes' to 'no' to disable
     --collapse: yes
     --trimns: yes
     --trimqualities: yes


# Mapping targets are specified using the following structure. Uncomment and
# replace 'NAME_OF_TARGET' with the desired prefix for filenames.
#NAME_OF_TARGET:
   #  Uncomment and replace 'NAME_OF_SAMPLE' with the name of this sample.
#  NAME_OF_SAMPLE:
     #  Uncomment and replace 'NAME_OF_LIBRARY' with the name of this sample.
#    NAME_OF_LIBRARY:
       # Uncomment and replace 'NAME_OF_LANE' with the name of this lane,
 # and replace 'PATH_WITH_WILDCARDS' with the path to the FASTQ files
       # to be trimmed and mapped for this lane (may include wildcards).
#      NAME_OF_LANE: PATH_WITH_WILDCARDS


D15_possort_20190213:
  D15_possort_20190213:
    D15_possort_20190213:
      "D15_possort_20190213": fastq/HFFposTH2AmCherry_D15_possort_20190213_S11_R1_001.fastq.gz
D15_possort_20190228:
  D15_possort_20190228:
    D15_possort_20190228:
      "D15_possort_20190228": fastq/HFFposTH2AmCherry_D15_possort_20190228_S13_R1_001.fastq.gz
...
```
Save and close the file.
Then, for running paleomix:

```{r}
paleomix trim_pipeline run --adapterremoval-max-threads=4 --max-threads=24 paleo_trim.yaml
```

#FastQC of the trimmed files

After paleomix, we need to continue working with the **truncated.gz** files. Files from paleomix analysis can be extracted as follows (one folder is created per sample):

```{r}
find -name "*truncated.gz" | awk '{split($1, path, "/"); system("mv "$1 " "path[2] "_" path[7])}'
```

FastQC can be run now:

```{r}
ls *truncated.gz | parallel -j 8 "fastqc --nogroup --outdir=  {}" &
```

Again, to create a report with the summary of all samples, use multiqc:

```{r}
multiqc #outdirs
```


#SortMeRNA for the removal of rRNA 

SortMeRNA was used for the removal of rRNA from the data (https://github.com/biocore/sortmerna). Before using this tool, rRNA database needs to be downloaded and indexed (git clone https://github.com/biocore/sortmerna.git). Then for running the tool:

```{r}
for i in `find . -name "*truncated.gz"`

do

sortmerna --ref \
~/sortmerna/rRNA_databases/silva-bac-16s-id90.fasta,~/sortmerna/idx/silva-bac-16s-id90:\
~/sortmerna/rRNA_databases/silva-bac-23s-id98.fasta,~/sortmerna/idx/silva-bac-23s-id98:\
~/sortmerna/rRNA_databases/silva-arc-16s-id95.fasta,~/sortmerna/idx/silva-arc-16s-id95:\
~/sortmerna/rRNA_databases/silva-arc-23s-id98.fasta,~/sortmerna/idx/silva-arc-23s-id98:\
~/sortmerna/rRNA_databases/silva-euk-18s-id95.fasta,~/sortmerna/idx/silva-euk-18s-id95:\
~/sortmerna/rRNA_databases/silva-euk-28s-id98.fasta,~/sortmerna/idx/silva-euk-28s-id98:\
~/sortmerna/rRNA_databases/rfam-5s-database-id98.fasta,~/sortmerna/idx/rfam-5s-database-id98:\
~/sortmerna/rRNA_databases/rfam-5.8s-database-id98.fasta,~/sortmerna/idx/rfam-5.8s-database-id98 \
-d ${i} \
--reads-gz $i --aligned /02_readswrRNA/${i}_rRNA --other /02_readsworRNA/${i}_no_rRNA --log -a 12 -v --fastx

done
```

#FastQC for SortMeRNA output

Quality control was performed again with the new fastq files obtained

```{r}
ls 02_readsworRNA/*.fastq | parallel -j 8 "fastqc --nogroup --outdir=  {}" &
  
multiqc #outdir
```

#Aligment of a reference genome using STAR

For the aligment of read to a reference genome, STAR was used (https://github.com/alexdobin/STAR). The following annotation files were used for aligment:

```{r}
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/gencode.v31.annotation.gff3.gz
gunzip gencode.v31.annotation.gff3.gz
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/GRCh38.p12.genome.fa.gz
gunzip GRCh38.p12.genome.fa.gz
```

Files need to be indexed:

```{r}
STAR-2.5.2b --runMode genomeGenerate --runThreadN 12 \
--genomeDir  \
--sjdbGTFfile ~/gencode.v31.annotation.gff3 \
--sjdbOverhang 74 \
--genomeFastaFiles ~/GRCh38.p12.genome.fa
```

Then, for running STAR:

```{r}
for i in *.fastq

do

STEM=$(basename "${i}" truncated.gz_no_rRNA.fastq)

STAR-2.5.2b --genomeDir  --runThreadN 8 \
--readFilesIn ${i} \
--outFileNamePrefix ~/"${STEM}" \
--twopassMode Basic --outSAMunmapped Within \
--limitOutSJcollapsed 1000000 --limitSjdbInsertNsj 1000000 --outFilterMultimapNmax 100 --outFilterMismatchNmax 33 \
--outFilterMismatchNoverLmax 0.3 --seedSearchStartLmax 12 --alignSJoverhangMin 15 --alignEndsType Local \
--outFilterMatchNminOverLread 0 --outFilterScoreMinOverLread 0.3 --winAnchorMultimapNmax 50 \
--alignSJDBoverhangMin 3 --outFilterType BySJout --outSAMtype BAM SortedByCoordinate

done
```

.bam files will be produced here
