---
title: "RNA-seq analysis. Time course differentiation data"
author: "Borja Gomez"
date: "2022-08-23"
output: html_document
---

##Introduction

Here is contained all the code used for the analysis of the time course differentiation data, including: smNPC, mDAN at days 15/30/50, non-mDAN at days 15/50, and astrocytes at day 65 of differentiation. Each sample contained a total of 3 replicates.

#Version of the different tools used

```{r}
samtools 1.9
Adapter Remval 2.2.2
bwa 0.7.17
fastqc 0.11.7
picard 2.18.13
paleomix 1.2.13.2
STAR 2.5.2b
CMake 3.12.1-intel-2018a
sortmerna 3.0.3
```

#Sample quality control by FastQC and MultiQC 

The FastQC tool (https://github.com/s-andrews/FastQC) was used to determine the quality of the RNA-seq samples obtained:

```{r}
ls *fastq.gz | parallel -j 8 "fastqc --nogroup --outdir=  {}" &
```

Then, in order to obtain all the results in a single report, MultiQC was used (https://multiqc.info/):

```{r}
multiqc #outdir
```

#Adapter trimming using Paleomix

For sequencing adapter removal, PALEOMIX was used (https://paleomix.readthedocs.io/en/stable/):

```{r}
#Create makefile 
paleomix trim_pipeline mkfile > paleo_trim.yaml

#Open the .yaml file to modify it
vim paleo_trim.yaml
```

Once inside the paleo_trim.yaml file, some changes were added to the file plus the path and short name for the fastq files was specified at the end (2 examles are shown):

```{r}
# -*- mode: Yaml; -*-
# Timestamp: 2019-07-18T10:05:18.882102
#
# Default options.
# Can also be specific for a set of samples, libraries, and lanes,
# by including the "Options" hierarchy at the same level as those
# samples, libraries, or lanes below. This does not include
# "Features", which may only be specific globally.
Options:
  # Sequencing platform, see SAM/BAM reference for valid values
  Platform: Illumina
  # Quality offset for Phred scores, either 33 (Sanger/Illumina 1.8+)
  # or 64 (Illumina 1.3+ / 1.5+). For Bowtie2 it is also possible to
  # specify 'Solexa', to handle reads on the Solexa scale. This is
  # used during adapter-trimming and sequence alignment
  QualityOffset: 33
  # Split a lane into multiple entries, one for each (pair of) file(s)
  # found using the search-string specified for a given lane. Each
  # lane is named by adding a number to the end of the given barcode.
  SplitLanesByFilenames: yes
  # Compression format for FASTQ reads; 'gz' for GZip, 'bz2' for BZip2
  CompressionFormat: gz

  # Settings for trimming of reads, see AdapterRemoval man-page
  AdapterRemoval:
     # Adapter sequences, set and uncomment to override defaults
#     --adapter1: AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC
#     --adapter2: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT
     # Some BAM pipeline defaults differ from AR defaults;
     # To override, change these value(s):
 --mm: 3
     --minlength: 25
     # Extra features enabled by default; change 'yes' to 'no' to disable
     --collapse: yes
     --trimns: yes
     --trimqualities: yes


# Mapping targets are specified using the following structure. Uncomment and
# replace 'NAME_OF_TARGET' with the desired prefix for filenames.
#NAME_OF_TARGET:
   #  Uncomment and replace 'NAME_OF_SAMPLE' with the name of this sample.
#  NAME_OF_SAMPLE:
     #  Uncomment and replace 'NAME_OF_LIBRARY' with the name of this sample.
#    NAME_OF_LIBRARY:
       # Uncomment and replace 'NAME_OF_LANE' with the name of this lane,
 # and replace 'PATH_WITH_WILDCARDS' with the path to the FASTQ files
       # to be trimmed and mapped for this lane (may include wildcards).
#      NAME_OF_LANE: PATH_WITH_WILDCARDS


D15_possort_20190213:
  D15_possort_20190213:
    D15_possort_20190213:
      "D15_possort_20190213": fastq/HFFposTH2AmCherry_D15_possort_20190213_S11_R1_001.fastq.gz
D15_possort_20190228:
  D15_possort_20190228:
    D15_possort_20190228:
      "D15_possort_20190228": fastq/HFFposTH2AmCherry_D15_possort_20190228_S13_R1_001.fastq.gz
...
```
Save and close the file.
Then, for running paleomix:

```{r}
paleomix trim_pipeline run --adapterremoval-max-threads=4 --max-threads=24 paleo_trim.yaml
```

#FastQC of the trimmed files

After paleomix, we need to continue working with the **truncated.gz** files. Files from paleomix analysis can be extracted as follows (one folder is created per sample):

```{r}
find -name "*truncated.gz" | awk '{split($1, path, "/"); system("mv "$1 " "path[2] "_" path[7])}'
```

FastQC can be run now:

```{r}
ls *truncated.gz | parallel -j 8 "fastqc --nogroup --outdir=  {}" &
```

Again, to create a report with the summary of all samples, use multiqc:

```{r}
multiqc #outdirs
```


#SortMeRNA for the removal of rRNA 

SortMeRNA was used for the removal of rRNA from the data (https://github.com/biocore/sortmerna). Before using this tool, rRNA database needs to be downloaded and indexed (git clone https://github.com/biocore/sortmerna.git). Then for running the tool:

```{r}
for i in `find . -name "*truncated.gz"`

do

sortmerna --ref \
~/sortmerna/rRNA_databases/silva-bac-16s-id90.fasta,~/sortmerna/idx/silva-bac-16s-id90:\
~/sortmerna/rRNA_databases/silva-bac-23s-id98.fasta,~/sortmerna/idx/silva-bac-23s-id98:\
~/sortmerna/rRNA_databases/silva-arc-16s-id95.fasta,~/sortmerna/idx/silva-arc-16s-id95:\
~/sortmerna/rRNA_databases/silva-arc-23s-id98.fasta,~/sortmerna/idx/silva-arc-23s-id98:\
~/sortmerna/rRNA_databases/silva-euk-18s-id95.fasta,~/sortmerna/idx/silva-euk-18s-id95:\
~/sortmerna/rRNA_databases/silva-euk-28s-id98.fasta,~/sortmerna/idx/silva-euk-28s-id98:\
~/sortmerna/rRNA_databases/rfam-5s-database-id98.fasta,~/sortmerna/idx/rfam-5s-database-id98:\
~/sortmerna/rRNA_databases/rfam-5.8s-database-id98.fasta,~/sortmerna/idx/rfam-5.8s-database-id98 \
-d ${i} \
--reads-gz $i --aligned /02_readswrRNA/${i}_rRNA --other /02_readsworRNA/${i}_no_rRNA --log -a 12 -v --fastx

done
```

#FastQC for SortMeRNA output

Quality control was performed again with the new fastq files obtained

```{r}
ls 02_readsworRNA/*.fastq | parallel -j 8 "fastqc --nogroup --outdir=  {}" &
  
multiqc #outdir
```

#Aligment of a reference genome using STAR

For the aligment of read to a reference genome, STAR was used (https://github.com/alexdobin/STAR). The following annotation files were used for aligment:

```{r}
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/gencode.v31.annotation.gff3.gz
gunzip gencode.v31.annotation.gff3.gz
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/GRCh38.p12.genome.fa.gz
gunzip GRCh38.p12.genome.fa.gz
```

Files need to be indexed:

```{r}
STAR-2.5.2b --runMode genomeGenerate --runThreadN 12 \
--genomeDir  \
--sjdbGTFfile ~/gencode.v31.annotation.gff3 \
--sjdbOverhang 74 \
--genomeFastaFiles ~/GRCh38.p12.genome.fa
```

Then, for running STAR:

```{r}
for i in *.fastq

do

STEM=$(basename "${i}" truncated.gz_no_rRNA.fastq)

STAR-2.5.2b --genomeDir  --runThreadN 8 \
--readFilesIn ${i} \
--outFileNamePrefix ~/"${STEM}" \
--twopassMode Basic --outSAMunmapped Within \
--limitOutSJcollapsed 1000000 --limitSjdbInsertNsj 1000000 --outFilterMultimapNmax 100 --outFilterMismatchNmax 33 \
--outFilterMismatchNoverLmax 0.3 --seedSearchStartLmax 12 --alignSJoverhangMin 15 --alignEndsType Local \
--outFilterMatchNminOverLread 0 --outFilterScoreMinOverLread 0.3 --winAnchorMultimapNmax 50 \
--alignSJDBoverhangMin 3 --outFilterType BySJout --outSAMtype BAM SortedByCoordinate

done
```

.bam files were produced here 

#Q30 filtering plus validation of bam files 

Samtools was used for Q30 filtering of quality reads and reduce the possibility of having false positives in the data (https://github.com/samtools/samtools):

```{r}
parallel -j 8 "samtools view -b -q 30 {} > ~/"{.}".q30.bam" ::: *.out.bam
```

Then, Picard was used for the validation of the bam files (https://github.com/broadinstitute/picard):

```{r}
#For running this, Java needs to be loaded

parallel -j 6 "java -jar ~/jar_root/picard.jar ValidateSamFile I={} O={.}.validated IGNORE=MISSING_TAG_NM IGNORE=MISSING_READ_GROUP IGNORE=READ_GROUP_NOT_FOUND IGNORE_WARNINGS=TRUE" ::: *.Aligned.sortedByCoord.out.bam

```

Picard will produce a .validated file. If everything was right, if we open the file it should say No errors found.

#Obtaining gene counts using FeatureCounts from Rsubread

For obtaining the gene counts, the following packages were needed:

```{r}
#R version used was 3.4.4
install.packages("tidyverse") 

source("https://bioconductor.org/biocLite.R")
biocLite("Rsubread") 

```

Then:

```{r}
bams <- Sys.glob("*q30.bam")
annot <- "~/gencode.v31.annotation.gff3"
fc_hs <- Rsubread::featureCounts(files = bams, annot.ext = annot, nthreads = 8, 
   isGTFAnnotationFile = TRUE, GTF.featureType = "exon", GTF.attrType = "gene_id", 
   strandSpecific = 2)

saveRDS(fc_hs, "fc_hs_q30RNA.rds")
```

.rds counts contain the counts per sample


#DESeq2 analysis 

To determine gene expression changes between the different samples analyzed, DESeq2 package from R was used (https://bioconductor.org/packages/release/bioc/html/DESeq2.html). The following packages had to be loaded:

```{r}
library(tidyverse)
library(DESeq2)
library(knitr)
library(rtracklayer)
library(RColorBrewer)
library(pheatmap)
```

Extract information for the .rds file:

```{r}
fc_hs <- readr::read_rds("~/fc_hs_q30RNA.rds")
rawcounts <- fc_hs$counts
```

A metadate file had to be created in order to run the analysis. Here is an example of what was used:

```{r}
sample <- c("Astro", "Astro", "Astro", "D15_NEG", "D15_NEG", "D15_NEG", "D15_POS", "D15_POS", "D15_POS", "D30_POS", "D30_POS", "D30_POS", "D50_NEG", "D50_NEG", "D50_NEG", "D50_POS", "D50_POS", "D50_POS", "smNPC", "smNPC", "smNPC")
cell_type <- c("astrocyte", "astrocyte", "astrocyte", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "neuron", "smNPC", "smNPC", "smNPC")
age <- c("65", "65", "65", "15", "15", "15", "15", "15", "15", "30", "30", "30", "50", "50", "50", "50", "50", "50", NA, NA, NA)
sort <- c(NA, NA, NA, "neg", "neg", "neg", "pos", "pos", "pos", "pos", "pos", "pos", "neg", "neg", "neg", "pos", "pos", "pos", NA, NA, NA)

info.data <- data.frame(sample, cell_type, age, sort)
```

Then, DESeq2 object was created:

```{r}
dds <- DESeqDataSetFromMatrix(countData = rawcounts,
                              colData = info.data,
                              design = ~ sample)

dds$sample <- factor(dds$sample, levels = c("smNPC", "D15_NEG", "D15_POS", "D30_POS", "D50_NEG", "D50_POS", "Astro"))
```

For creating a PCA plot:

```{r}
vst <- vst(dds, blind = TRUE)

plotPCA(vst, intgroup = c("cell_type", "age", "sort"))
```

Low expressed genes were filtered from the dataset

```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```

After the filtering, DESeq2 analysis was run:

```{r}
dds <- DESeq(dds)
```

In order to be able to compare different groups, the package "ashr" was used:

```{r}
library(ashr)
#Here is an example for the comparison of smNPC with astrocytes, although it can be done among all groups
result.astro.LFC <- lfcShrink(dds, contrast = c("sample", "Astro", "smNPC"), type = "ashr", lfcThreshold = 1)
```

#RPKM calculation for plotting the data

Extract the rawcounts from the dds object. Remember that the dds object that we are using had a previous filtering:
```{r}
rawcount_2 <- counts(dds)
```
Extract the gene_length for the genes in rawcount_2 (for annotation use gencode.v31.annotation.gff3 file):

```{r}
gene_length <- fc_hs$annotation %>% 
  select(GeneID, Length) %>%
  filter(GeneID %in% rownames(rawcount_2))
   
```

```{r}
#After extraction of the gene length, we transform the results from bp to kb
gene_length <- gene_length %>% 
    mutate(length_kb = Length/1000)
  
# scaling factor. Here we are obtaining the reads per million for each individual sample. It is row wise. This is indicated using the number 2 in the function. 
  scale_factor <- apply(rawcount_2,
    2, function(x) sum(x))/1000000
  scale_factor
  
# normalize to library size. 
norm_counts <- rawcount_2 %>%  
    sweep(., 2, scale_factor, "/")
norm_counts
  
# rpkm. This is divided again by the gene length. 
rawcount_2 %>% as.data.frame() %>% 
  rownames_to_column(var = "GeneID") -> rawcount_2
all_info <- left_join(rawcount_2, gene_length, by = c("GeneID")) 
  
rpkm <- norm_counts %>% 
  sweep(., 1, all_info$length_kb, "/")

rpkm
```


#Calculating Z-score for producing heatmaps

```{r}
vsd <- assay(vst(dds))
Z-score <- t(scale(t(vsd)))
```

Then, genes of interest can be selected for plotting using pheatmap. Here an example:

```{r}
#Z-score-list would represent filtered list of selected genes. 
pheatmap(Z-score-list, annotation_col = select(info.data, cell, type), show_colnames = FALSE, cluster_cols = FALSE, filename = "Data/Heatmap_cholesterol_cells.pdf")
```

#Plotting individual genes 

For plotting individual genes, RPKM values were used but data had to be rearranged:

```{r}
left_join(rpkm,
          annotation,
          by = "GeneID") %>% distinct(gene_name, .keep_all = TRUE) -> rpkm

as.data.frame(t(rpkm)) %>% rownames_to_column(var = "samples") -> rpkm

rpkm$cell <- c("Astro", "Astro", "Astro", "D15_NEG", "D15_NEG", "D15_NEG", "D15_POS", "D15_POS", "D15_POS", "D30_POS", "D30_POS", "D30_POS", "D50_NEG", "D50_NEG", "D50_NEG", "D50_POS", "D50_POS", "D50_POS", "smNPC", "smNPC", "smNPC")

rpkm$sort <- c(NA, NA, NA, "neg", "neg", "neg", "pos", "pos", "pos", "pos", "pos", "pos", "neg", "neg", "neg", "pos", "pos", "pos", NA, NA, NA)

rpkm$cell <- factor(rpkm$cell, levels = c("smNPC", "D15_NEG", "D15_POS", "D30_POS", "D50_NEG", "D50_POS", "Astro"))
```

For plotting an individual gene:

```{r}
ggplot(rpkm) + 
  geom_boxplot(aes(x=cell, y= gene_name, fill = sort)) +
  geom_jitter(aes(x=cell, y= gene_name), alpha = 1) +
  theme_light() +
  labs(title = "",
       x = "",
       y = "RPKM") +
  theme(plot.title = element_text(size = 20, face = "bold", hjust = 0.5), axis.title.y = element_text(face = "bold")) + expand_limits(y = 0) -> Gene_expression

ggsave(plot = Gene_expression, filename = "Data/Gene_expression.pdf", dpi = 300)
```














